Remotion/Web/ClientScript/Scripts/SmartNavigation.ts

```diff
@@ -0,0 +1,249 @@
+  {
+    if (scrollElement == null)
+      return;
+    var htmlElement = window.document.getElementById (scrollElement.ID);
+    if (htmlElement == null)
+      return;
+    htmlElement.scrollTop = scrollElement.Top;
+    htmlElement.scrollLeft = scrollElement.Left;
+  }
+}
+
+function SmartScrolling_Restore (data: Nullable<string>): void
+{
+  SmartScrolling.Restore (data);
+}
+
+function SmartScrolling_Backup(): string
+{
+  return SmartScrolling.Backup();
+}
+
+function SmartScrolling_GetScrollPositions (currentElement: Nullable<HTMLElement>): SmartScrolling_Element[]
+{
+  return SmartScrolling.GetScrollPositions (currentElement);
+}
+
+function SmartScrolling_GetScrollPosition (htmlElement: Nullable<HTMLElement>): Nullable<SmartScrolling_Element>
+{
+  return SmartScrolling.GetScrollPosition (htmlElement);
+}
+
+function SmartScrolling_SetScrollPosition (scrollElement: Nullable<SmartScrolling_Element>): void
+{
+  return SmartScrolling.SetScrollPosition (scrollElement);
+}
+
+class SmartFocus
+{
+  public static Backup(): string
+  {
+    var data = '';
+    var activeElement = window.document.activeElement;
+    if (activeElement != null)
+    {
+      data += activeElement.id;
+    }
+    return data;
+  }
+
+  public static Restore (data: string): boolean
+  {
+    var activeElementID = data;
+    if (! StringUtility.IsNullOrEmpty (activeElementID))
+    {
+      var activeElement = $('#' + activeElementID);
+      if (activeElement)
+      {
+        var isEnabledFilter = function (this: HTMLElement)
+        {
+          var _this = $ (this);
+          if (_this.is ('button'))
+            return true;
+          if (_this.is ('input[type=button]'))
+            return true;
+          if (_this.is ('input[type=submit]'))
+            return true;
+          if (_this.is ('a'))
+            return true;
+          if (!_this.is ('a') && _this.is (':enabled') && !_this.is ('input[type=hidden]'))
+            return true;
+          if (!_this.is ('a') && _this.is ('*[tabindex]'))
+            return true;
+          return false;
+        };
+
+        if (activeElement.is(isEnabledFilter))
+        {
+          SmartFocus.SetFocus (activeElement);
+        }
+        else
+        {
+          var focusableElements = $('input, textarea, select, button, a');
+          var elementIndex = focusableElements.index(activeElement);
+          var fallBackElement = focusableElements.slice(elementIndex).filter(isEnabledFilter).first();
+          if (fallBackElement.length > 0)
+          {
+            SmartFocus.SetFocus (fallBackElement);
+          }
+          else
+          {
+            fallBackElement = focusableElements.slice(0, elementIndex).filter(isEnabledFilter).last();
+            if (fallBackElement)
+            {
+              SmartFocus.SetFocus (fallBackElement);
+            }
+          }
+        }
+      }
+    }
+    // TODO: implicitly returns undefined
+    return false;
```

Comment:
@MichaelKetting do you want to fix this issue in a separate issue?

Reply:
Ah, yes, that should be a separate ticket. I didn't think of this at first but there's the slight possiblity that the undefined-case results in different behavior from the false-case and I wouldn't want that change hidden in a TypeScript migration.